proof/refine/AARCH64/Schedule_R.thy

```diff
@@ -156,16 +156,127 @@ lemma vcpu_at_cross:
+  "corres dc (pspace_aligned and pspace_distinct and valid_cur_fpu and none_top tcb_at new_owner) \<top>
+             (set_arm_current_fpu_owner new_owner) (modifyArchState (armKSCurFPUOwner_update (\<lambda>_. new_owner)))"
+  unfolding set_arm_current_fpu_owner_def modifyArchState_def maybeM_def
+  apply corres_pre
+  apply (rule corres_underlying_gets_pre_lhs)
+  apply (rule corres_add_noop_rhs)
+  apply (corres_split; clarsimp?)
+       apply (corres_cases; corres corres: set_tcb_cur_fpu_noop)
+      apply (rule corres_add_noop_rhs2)
+      apply (rule corres_split)
+         apply (corres corres: corres_modify_tivial) \<comment> \<open>FIXME: typo in rule name\<close>
+         apply (clarsimp simp: state_relation_def arch_state_relation_def)
+        apply (corres_cases; corres corres: set_tcb_cur_fpu_noop)
+       apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift)+
+   apply (auto simp: current_fpu_owner_Some_tcb_at)
+  done
+
+lemma enableFpu_fpu_enabled[wp]:
+  "\<lbrace>\<top>\<rbrace> enableFpu \<lbrace>\<lambda>_. fpu_enabled\<rbrace>"
+  by (wpsimp simp: enableFpu_def)
+
+crunch writeFpuState, readFpuState
+  for fpu_enabled[wp]: fpu_enabled
+
+crunch load_fpu_state, save_fpu_state
+  for pspace_aligned[wp]: pspace_aligned
+  and pspace_distinct[wp]: pspace_distinct
+  and valid_cur_fpu[wp]: valid_cur_fpu
+  and fpu_enabled[wp]: "\<lambda>s. fpu_enabled (machine_state s)"
+  (wp: dmo_machine_state_lift)
+
+defs fpuOwner_asrt_def:
+  "fpuOwner_asrt \<equiv> \<lambda>s'. opt_tcb_at' (armKSCurFPUOwner (ksArchState s')) s'"
+
+lemma fpuOwner_asrt_cross:
+  "\<lbrakk>(s, s') \<in> state_relation; valid_cur_fpu s; pspace_aligned s; pspace_distinct s\<rbrakk> \<Longrightarrow> fpuOwner_asrt s'"
+  by (fastforce simp: state_relation_def arch_state_relation_def fpuOwner_asrt_def
+              intro!: tcb_at_cross current_fpu_owner_Some_tcb_at)
+
+lemma switchLocalFpuOwner_corres[corres]:
+  "corres dc (pspace_aligned and pspace_distinct and valid_cur_fpu and none_top tcb_at new_owner) \<top>
+             (switch_local_fpu_owner new_owner) (switchLocalFpuOwner new_owner)"
+  unfolding switch_local_fpu_owner_def switchLocalFpuOwner_def maybeM_def
+  apply (corres corres: corres_stateAssert_r | corres_cases | clarsimp)+
+         apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift' dmo_machine_state_lift)+
+   apply (auto simp: current_fpu_owner_Some_tcb_at fpuOwner_asrt_cross)
+  done
+
+lemma isFlagSet_set:
+  "isFlagSet flag flags = (flag \<in> word_to_tcb_flags flags)"
+  by (clarsimp simp: isFlagSet_def word_to_tcb_flags_def)
+
+lemma lazyFpuRestore_corres[corres]:
+  "corres dc (pspace_aligned and pspace_distinct and valid_cur_fpu and tcb_at t) \<top>
+             (lazy_fpu_restore t) (lazyFpuRestore t)"
+  unfolding lazy_fpu_restore_def lazyFpuRestore_def
+  by (corres corres: threadGet_corres[where r="\<lambda>flags flags'. flags = word_to_tcb_flags flags'"] \<comment> \<open>FIXME: make a tcb_flags_relation?\<close>
```

Comment:
> FIXME: make a tcb_flags_relation?

Unless there is a lot more of the same further below, it's probably not worth making a separate definition for it.